### PR TITLE
[SE-0290] Add #unavailable (Negative platform checks)

### PR DESCRIPTION
**Pull request by rockbruno**
Implementation for SE-0290: https://github.com/apple/swift-evolution/blob/main/proposals/0290-negative-availability.md

Swift evolution thread: https://forums.swift.org/t/support-negative-availability-literals/

### CODE REVIEW

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

**Comment by rockbruno** [comment]
I notice now that that the underlying `_stdlib_isOSVersionAtLeast` emission needs to be reverted as well. I'll add this change along with some SIL tests as the review progresses.

**Comment by rockbruno** [comment]
I updated the implementation to match the final revision (imply the wildcard). The implementation now rejects explicit `*`s and injecting a fake one to the spec list so that the rest of the logic works as expected.

I was unsure how to proceed with `#unavailable()` though. This is technically correct if the wildcard is implied but right now it's being rejected by the parser.

File: lib/SILGen/SILGenDecl.cpp
Comment (rockbruno) [inline lib/SILGen/SILGenDecl.cpp:1356]: I was unsure if this is the best way of inverting the result. IIRC the `!` operator XORs the result, which should be faster but I couldn't figure out to emit a call to it here.

**Comment by rockbruno** [comment]
@tkremenek Any updates on the plan for this feature? I'd like to keep the branch updated, but only when this actually moves forward.

File: lib/Parse/ParseStmt.cpp
Comment (varungandhi-apple) [inline lib/Parse/ParseStmt.cpp:1307]: Should this replacement be `!#unavailable`? It seems a bit strange for `#available` to be replaced by `#unavailable` since they have roughly the opposite meaning.

File: include/swift/AST/Stmt.h
Comment (varungandhi-apple) [inline include/swift/AST/Stmt.h:389]: I think it would be better to rename the method to `isUnavailability` and rename the field to `_isUnavailability` (or something else). Predicate methods should generally be named `isXYZ()`.

File: include/swift/AST/DiagnosticsParse.def
Comment (varungandhi-apple) [inline include/swift/AST/DiagnosticsParse.def:1799]: I think it would be slightly better to rephrase this to something like:

```
platform wildcard '*' is always implicit in '#unavailable'
```

The "always" indicates that it is (erm) always implicit, never explicit. Also "unavailability" is a mouthful of a word (it's 7 syllables), other diagnostics are using `'#unavailable'` (5 syllables) so IMO it makes sense to be consistent.

File: lib/SILGen/SILGenDecl.cpp
Comment (varungandhi-apple) [reply lib/SILGen/SILGenDecl.cpp:1356]: I don't know if there's a better way to do it but you can use `createIntegerLiteral(loc, i1, 1);` and `xor` instead of `cmp_eq`, I think that should do the trick.

**Comment by varungandhi-apple** [comment]
Sorry this got missed. Could you please fix the merge conflicts? I've left some comments and have asked other folks to chime in as well.

In the future, please feel free to ping someone who has modified the surrounding code for review, or ask on the Swift forums for review.

**Comment by varungandhi-apple** [comment]
cc @rintaro for parser changes, cc @amartini51 as there may be potential TSPL changes.

File: test/Parse/availability_query_unavailablility.swift
Comment (varungandhi-apple) [inline test/Parse/availability_query_unavailablility.swift:100]: You can remove this line, this has since been fixed. 😃 Maybe also remove the corresponding line from the original file.

File: lib/Parse/ParseStmt.cpp
Comment (varungandhi-apple) [reply lib/Parse/ParseStmt.cpp:1307]: Sorry, I misunderstood the fix-it, the `== false` is getting removed as well. This seems correct.

File: lib/Parse/ParseStmt.cpp
Comment (varungandhi-apple) [inline lib/Parse/ParseStmt.cpp:1469]: Nit: formatting is off. `peekToken().getLoc()` should probably go on the earlier line.

**Comment by rockbruno** [comment]
Thanks @varungandhi-apple! I'll rebase it and apply the changes.

File: lib/SILGen/SILGenDecl.cpp
Comment (rockbruno) [reply lib/SILGen/SILGenDecl.cpp:1356]: I got it now! I don't remember what the problem was before, I probably just confused it with something else 😄

**Comment by rockbruno** [comment]
@varungandhi-apple I've updated it. Had to lose the commit history though because it was 3000 commits behind main...

File: include/swift/AST/DiagnosticsParse.def
Comment (varungandhi-apple) [inline include/swift/AST/DiagnosticsParse.def:1822]: Nit/Question: I just realized, that other diagnostics are not putting `#available` in quotes, so should we remove the quotes here?

File: include/swift/Parse/Parser.h
Comment (varungandhi-apple) [inline include/swift/Parse/Parser.h:1758]: Nit: This function has 3 call-sites; I think it's best to avoid defaulting this argument, so that call-sites are clearer and can be marked with `/*isUnavailability*/value`.

File: include/swift/Parse/Parser.h
Comment (varungandhi-apple) [inline include/swift/Parse/Parser.h:1787]: An availability macro should serve the dual purpose for working both with `#available` and `#unavailable`. The availability spec list for an availability macro behaves like the spec list for `#unavailable` (see usage of `@available(SwiftStdlib 5.5, *)` in the stdlib; the definition of SwiftStdlib from compiler flags doesn't include `*`). That said, there's no fix-it in that case. Instead of having a pair of booleans (`isUnavailability` and isParsingMacroDefinition`) in multiple places, which indicates four possibilities, I think it would be best to try to use a dedicated enum with 3 cases. Something like:


```
enum class AvailabilitySpecSource: uint8_t {
  Available,
  Unavailable,
  Macro
}

void printAvailabilitySpecSource(llvm::raw_ostream &os, AvailabilitySpecSource source) { ... }
void dumpAvailabilitySpecSource(AvailabilitySpecSource source) { /* call print with llvm::errs() */ }
```

Then the body of this function would pass the `Macro` case to `parseAvailabilitySpecList`. Other places would pass `Available` or `Unavailable` as applicable.

**Comment by varungandhi-apple** [comment]
Please add some tests for the `#unavailable` using the result of expanding an availability macros in `Sema/availability_define.swift` (I think that's the only file where tests are needed for semantic checks but cc @xymus in case I missed something).

File: include/swift/Parse/Parser.h
Comment (varungandhi-apple) [reply include/swift/Parse/Parser.h:1758]: Nevermind, my later comment makes this redundant.

File: lib/Parse/ParseStmt.cpp
Comment (varungandhi-apple) [inline lib/Parse/ParseStmt.cpp:1246]: There is a minor bug-fix opportunity here as well as opportunity to make the code clearer.

The bug is that with an availability macro, this code is assuming that `*` is the last entry (which is very likely but not strictly guaranteed). So if you make a command like:

```
xcrun swiftc -Xfrontend -define-availability -Xfrontend "SwiftStdlib 5.5:macOS 9999, *,
watchOS 9999, tvOS 9999" tmp.swift
-define-availability argument:1:51: error: future platforms identified by '*' cannot be used in an availability macro definition
SwiftStdlib 5.5:macOS 9999, *, watchOS 9999, tvOS 9999
                                                  ^
```
the caret is pointing to the wrong token. Instead that check should be happening inside the loop similar to where the unavailable + `*` check is happening right now.

The Unavailabile and Macro cases are similar in that we need to diagnose an error if a wildcard is present. The difference is that in the Unavailable case, you need to provide a fix-it as well.

Here, the nested-if makes things a bit trickier to follow IMO. It would be nicer to have a switch:

```
if (!HasOtherPlatformSpec) { 
  switch (specSource) {
  case AvailabilitySpecSource::Macro: break;
  case AvailabilitySpecSource::Available: /* issue error and fix-it */
  case AvailabilitySpecSource::Unavailable: /* store new spec */
  }
}
```

**Comment by varungandhi-apple** [comment]
No worries about merging the commits. I think as a feature, it seems like it fits okay in one commit since it touches lots of things which are connected. Please do add a link to the proposal in the commit message though.

File: include/swift/Parse/Parser.h
Comment (rockbruno) [reply include/swift/Parse/Parser.h:1787]: Aha, now I understand what this function is for. I wasn't aware this was for the command line feature. Makes sense to configure it as an enum now.

Comment (rockbruno) [reply include/swift/Parse/Parser.h]: @varungandhi-apple I updated the branch and fixed the bug with the macro. With the enum, I was also able to clean up some of the logic as well 👍 

I just don't understand what you meant with `printAvailabilitySpecSource` and `dumpAvailabilitySpecSource` though, I saw some types with similar methods but I don't enough about the compiler to know what they are used for. Could you clarify that?

Comment (varungandhi-apple) [reply include/swift/Parse/Parser.h]: I suggested adding them to help with debugging, since C++ doesn't have automatic printing facilities like Swift does.

File: test/attr/attr_availability_unavailability.swift
Comment (varungandhi-apple) [inline test/attr/attr_availability_unavailability.swift:3]: Nit: please use `macOS` instead of `OSX` here.

File: test/SILGen/availability_query.swift
Comment (varungandhi-apple) [inline test/SILGen/availability_query.swift:119]: Not sure why there is a FIXME here... Is it a bug that this warning is reported? (If so, please file a bug report and not the bug report number here. Otherwise, you can remove the FIXME)

File: lib/Parse/ParseStmt.cpp
Comment (rintaro) [inline lib/Parse/ParseStmt.cpp:1333]: Could we actually consider this as `#unavailable`? In other words, could we `isUnavailability = true` here?

File: lib/Parse/ParseStmt.cpp
Comment (rintaro) [inline lib/Parse/ParseStmt.cpp:1255]: This is semantic thing, not in the source code. Could we move this out of `Parse`?
Also, Using `SourceLoc()` here means [`PoundAvailableInfo::getEndLoc()` may return invalid `SourceLoc`](https://github.com/apple/swift/blob/4cc1035ba22a4d0726e3a00f7dad75463d54e5de/lib/AST/Stmt.cpp#L295-L302), that means `PoundAvailableInfo::getSourceRange()` may hit assertion failure. Not sure this is actually reachable, but still.

File: lib/Parse/ParseStmt.cpp
Comment (rockbruno) [reply lib/Parse/ParseStmt.cpp:1333]: It depends. The thing is that we can only detect this after parsing the list, and the list itself is parsed differently depending on what the token is (ignoring lacking wildcard). It would work if peeked the entire expression before parsing it, but I don't know if that would be a good idea. What do you think?

File: test/SILGen/availability_query.swift
Comment (rockbruno) [reply test/SILGen/availability_query.swift:119]: I think what they tried to say here is that checking for availability for a version equal or less than your minimum target should automatically resolve to a boolean like `*` does (resulting in that warning), but no such check exists today and someone could add it. Below this test there are similar cases but they expect no warnings, so I think it might be better to follow how it works today and expect nothing

File: lib/Parse/ParseStmt.cpp
Comment (rockbruno) [reply lib/Parse/ParseStmt.cpp]: Done, I'm picking the location of the latest query now too.

File: lib/AST/Stmt.cpp
Comment (rockbruno) [reply lib/AST/Stmt.cpp]: Ran the formatter in some files now, but the result changes a lot more stuff so I picked just the ones relevant to this PR.

File: lib/Parse/ParseStmt.cpp
Comment (rintaro) [reply lib/Parse/ParseStmt.cpp:1325]: Thanks for clarifying. I don't think it's worth it.

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Review by rintaro — COMMENTED** [review]
Thanks! Parser changes looks good to me. I defer to @varungandhi-apple

File: lib/Parse/ParseStmt.cpp
Comment (varungandhi-apple) [inline lib/Parse/ParseStmt.cpp:1227]: Style thing: it might be cleaner to have the `if`s inside the `switch` with early exits, and you can have short `if`s, like `if (OtherPlatformSpecLoc.hasValue()) { break; }`... this would reduce nesting from 2 levels to 1 for the logic. Even if you choose to not do that, please avoid using `default:` and spell out all other cases explicitly so that we get warnings when new cases are added.

File: lib/Parse/ParseStmt.cpp
Comment (varungandhi-apple) [inline lib/Parse/ParseStmt.cpp:1397]: Nit: Please avoid `default:`.

File: lib/Sema/BuilderTransform.cpp
Comment (varungandhi-apple) [inline lib/Sema/BuilderTransform.cpp:467]: Nit: Since only one caller is using the default argument, I think it would be better to make it mandatory.

File: lib/Sema/BuilderTransform.cpp
Comment (varungandhi-apple) [inline lib/Sema/BuilderTransform.cpp:482]: This seems like a bug; should it be the `endLoc` variable instead of `ifStmt->getThenStmt()->getEndLoc()`?

File: lib/Sema/BuilderTransform.cpp
Comment (varungandhi-apple) [inline lib/Sema/BuilderTransform.cpp:486]: Style nit: LLVM style recommends using early returns/exits, so you could `return arg;` after a check and unindent the other logic.

File: lib/Sema/TypeCheckAvailability.cpp
Comment (varungandhi-apple) [inline lib/Sema/TypeCheckAvailability.cpp:777]: Q: Why is this branch being removed?

File: lib/Sema/TypeCheckAvailability.cpp
Comment (varungandhi-apple) [inline lib/Sema/TypeCheckAvailability.cpp:810]:
```suggestion
    assert(isUnavailability.hasValue());
    if (isUnavailability.getValue()) {
      return std::make_pair(FalseRefinement, NestedTRC->getAvailabilityInfo());
    }
    return std::make_pair(NestedTRC->getAvailabilityInfo(), FalseRefinement);
```

(I would try to avoid the double `!` if possible.)

File: lib/Sema/BuilderTransform.cpp
Comment (rockbruno) [reply lib/Sema/BuilderTransform.cpp]: I think in fact that the whole implementation of the builder in this PR was confusing. I didn't know what I was changing in the beginning and I see now that it was over-complicated. I fixed that up and cleaned it up, though I'm not sure exactly where this loc is using to write a test for it, it seems to be only for diagnostics but they appear to the working fine even with the bug.

File: lib/Sema/TypeCheckAvailability.cpp
Comment (rockbruno) [reply lib/Sema/TypeCheckAvailability.cpp]: I inverted the position for them to become positive, but I think I can't add the assertion because this method can be called when there are no conditions as well (so `isUnavailability` would be None)

File: lib/Sema/TypeCheckAvailability.cpp
Comment (rockbruno) [reply lib/Sema/TypeCheckAvailability.cpp]: I see what you mean. The reason I removed it is just because I didn't want to duplicate the logic that inverts the result. It's not clear if there's a consequence of returning the current refinement instead of None, but I added it back with a closure instead.

File: lib/Sema/TypeCheckAvailability.cpp
Comment (varungandhi-apple) [inline lib/Sema/TypeCheckAvailability.cpp:799]: Nit: Please use C++ lambda syntax instead of using block syntax.

File: lib/Sema/TypeCheckAvailability.cpp
Comment (varungandhi-apple) [inline lib/Sema/TypeCheckAvailability.cpp:828]: It doesn't look like `Queries` is getting used later.
```suggestion
    for (auto *Spec : available->getQueries()) {
```

**Review by varungandhi-apple — APPROVED** [review]
Thanks for your patience with the review. Mostly LGTM, left a couple of minor comments. You may want to squash some of the later commits.

**Comment by varungandhi-apple** [comment]
@swift-ci smoke test

**Comment by rockbruno** [comment]
@varungandhi-apple There was a typo in the result builder logic, I cleaned it up further and squashed the rest.

**Comment by varungandhi-apple** [comment]
@swift-ci smoke test

**Comment by rockbruno** [comment]
Hm it failed because of SwiftSyntax not knowing the keyword, do I need to push something in its repo first for it to pass here?

**Comment by ahoppen** [comment]
> Hm it failed because of SwiftSyntax not knowing the keyword, do I need to push something in its repo first for it to pass here?

You need to re-generate the files in the SwiftSyntax repo. For this:
- Check out `https://github.com/apple/swift-syntax` next to your `swift` repository
- Run `./build-script.py --toolchain /path/to/anywhere --degyb-only` in the `swift-syntax` repository
- Set up a PR to the `swift-syntax` repository and mention it in this PR
- Not to people triggering CI: You need to reference the corresponding swift-syntax PR when testing swift and the swift PR when testing swift-syntax like [here](https://github.com/apple/swift-syntax/pull/281#issuecomment-829584400). Also the two PRs need to be merged at the same time.

**Comment by rockbruno** [comment]
@ahoppen Thanks! I tried that but I get errors about python imports. Do I need a specific python version?

```
    from .Classification import classification_by_name
ImportError: attempted relative import with no known parent package
```

**Comment by ahoppen** [comment]
I haven’t seen that error message before. I’m using Python 2. Could you share the entire error message? Maybe I’ll be able to figure out what’s going on from there.

**Comment by rockbruno** [comment]
With Python 2, I get a slightly different error:

```
** Generating gyb Files **
Traceback (most recent call last):
  File "/Users/rochab/Desktop/Other/swift-sources/swift/utils/gyb", line 3, in <module>
    gyb.main()
  File "/Users/rochab/Desktop/Other/swift-sources/swift/utils/gyb.py", line 1267, in main
    f.write(execute_template(ast, args.line_directive, **bindings))
  File "/Users/rochab/Desktop/Other/swift-sources/swift/utils/gyb.py", line 1137, in execute_template
    ast.execute(execution_context)
  File "/Users/rochab/Desktop/Other/swift-sources/swift/utils/gyb.py", line 639, in execute
    x.execute(context)
  File "/Users/rochab/Desktop/Other/swift-sources/swift/utils/gyb.py", line 725, in execute
    result = eval(self.code, context.local_bindings)
  File "/Users/rochab/Desktop/Other/swift-sources/swift-syntax/Sources/SwiftSyntaxBuilder/Buildables.swift.gyb", line 5, in <module>
    from gyb_syntax_support.kinds import syntax_buildable_child_type, syntax_buildable_default_init_value
ImportError: cannot import name syntax_buildable_default_init_value
FAIL: Generating .gyb files failed
```

**Comment by ahoppen** [comment]
That’s probably because `syntax_buildable_default_init_value` was added fairly recently (~1 month ago) in https://github.com/apple/swift/commit/b7d4dec54e87cfe093a9a0e966cd66c94b2f87eb. Could you try rebasing your swift branch?

**Comment by rockbruno** [comment]
Should've tried that first. It worked, thanks!

https://github.com/apple/swift-syntax/pull/296

**Comment by ahoppen** [comment]
Nice! Let’s see what CI has to say now.

https://github.com/apple/swift-syntax/pull/296
@swift-ci Please smoke test

**Review by ahoppen — APPROVED** [review]
Changes to `gyb_syntax_support` are looking good to me. I didn’t look at the other changes.

**Comment by varungandhi-apple** [comment]
@swift-ci test and merge

**Comment by rockbruno** [comment]
I think we need to merge https://github.com/apple/swift-syntax/pull/296 first, right?

**Comment by ahoppen** [comment]
You need to merge them at the same time. Make sure that both have a green CI status before merging either.

**Comment by rockbruno** [comment]
Ah I see, it's because we needed to reference the PR in the command.

**Comment by ahoppen** [comment]
Yeah, so they've always been tested in unison in PR testing so far (because you referenced the other PR) and thus you also need to merge them together.

**Comment by varungandhi-apple** [comment]
@rockbruno could you submit a PR adding a "Swift Next" entry to the ChangeLog and mentioning this new feature there? Here's an [example PR](https://github.com/apple/swift/commit/238290cdc4d15817ca09fc4bd31a18faf7c1328e) which does that.

**Comment by rockbruno** [comment]
Done: https://github.com/apple/swift/pull/38390

**Pull request merged**
